### PR TITLE
Rewrite existing repo units if setting is not included in api body

### PR DESCRIPTION
**Pull request by davidsvantesson**
Fix for #7758
In the API: PATCH ​/repos​/{owner}​/{repo}, if any advanced repo setting is not included in the body, corresponding repo_unit is rewritten (if existing) in order to retain the setting.

Signed-off-by: David Svantesson <davidsvantesson@gmail.com>

### CODE REVIEW

File: routers/api/v1/repo/repo.go [hunk 1]
```diff
@@ -651,7 +651,14 @@ func updateRepoUnits(ctx *context.APIContext, opts api.EditRepoOption) error {
...
 
-	if opts.HasIssues != nil {
+	if opts.HasIssues == nil {
+		// If HasIssues setting not touched, rewrite existing repo unit
+		if unit, err := repo.GetUnit(models.UnitTypeIssues); err == nil {
+			units = append(units, *unit)
+		} else if unit, err := repo.GetUnit(models.UnitTypeExternalTracker); err == nil {
+			units = append(units, *unit)
+		}
+	} else {
```
Comment (lafriks) [inline routers/api/v1/repo/repo.go]:
```suggestion
	} else if *opts.HasIssues {
```

File: routers/api/v1/repo/repo.go [hunk 2]
```diff
@@ -677,7 +684,14 @@ func updateRepoUnits(ctx *context.APIContext, opts api.EditRepoOption) error {
...
 
-	if opts.HasWiki != nil {
+	if opts.HasWiki == nil {
+		// If HasWiki setting not touched, rewrite existing repo unit
+		if unit, err := repo.GetUnit(models.UnitTypeWiki); err == nil {
+			units = append(units, *unit)
+		} else if unit, err := repo.GetUnit(models.UnitTypeExternalWiki); err == nil {
+			units = append(units, *unit)
+		}
+	} else {
```
Comment (lafriks) [inline routers/api/v1/repo/repo.go]: same here

File: routers/api/v1/repo/repo.go [hunk 3]
```diff
@@ -692,7 +706,12 @@ func updateRepoUnits(ctx *context.APIContext, opts api.EditRepoOption) error {
 		}
 	}
 
-	if opts.HasPullRequests != nil {
+	if opts.HasPullRequests == nil {
+		// If HasPullRequest setting not touched, rewrite existing repo unit
+		if unit, err := repo.GetUnit(models.UnitTypePullRequests); err == nil {
+			units = append(units, *unit)
+		}
+	} else {
```
Comment (lafriks) [inline routers/api/v1/repo/repo.go]: same here

**Review by lafriks — CHANGES_REQUESTED** [review]

File: routers/api/v1/repo/repo.go (hunk 1 above)
Comment (davidsvantesson) [reply routers/api/v1/repo/repo.go]: Fixed

File: routers/api/v1/repo/repo.go (hunk 2 above)
Comment (davidsvantesson) [reply routers/api/v1/repo/repo.go]: Fixed

File: routers/api/v1/repo/repo.go (hunk 3 above)
Comment (davidsvantesson) [reply routers/api/v1/repo/repo.go]: Fixed

**Review by lafriks — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Comment by davidsvantesson** [comment]
Ready to be merged?

**Comment by lafriks** [comment]
Please send backport to 1.9

**Pull request merged**